### DISCUSSION TITLE
ModuleNotFoundError: No module named 'spacy.gold'

### DISCUSSION THREAD

**Discussion by erotavlas** [explosion/spaCy#7660] [Help: Coding & Implementations] [answered]
Our annotation tool outputs annotations in this format (I'm not sure what this format is called in the documentation - but it was present in many examples)

`[("""Run date: 11/12/15 1705               """,{'entities':[(10,18,'DATE')]}),("""NAME: LNAMEZ,FNAME             DOB: 12/12/1901   """,{'entities':[(6,18,'PERSON'),(36,46,'DATE')]})]`

For Spacy versions prior to 3.0 I was using the following code to convert the above annotations to jsonl format like this

```
import os
import spacy
import srsly
from spacy.gold import docs_to_json, biluo_tags_from_offsets, spans_from_biluo_tags

path_train = r'D:\training\'
file_train = os.path.join(path_train, 'FINALTRAIN')

file = open(file_train, mode='r')
all_text = file.read()
file.close()

TRAIN_DATA = eval(all_text)

nlp = spacy.load(r'D:\spacy_baseModel\1.0.0.4') 

docs_train = []
for text, annot in TRAIN_DATA:
    doc = nlp(text)
    tags = biluo_tags_from_offsets(doc, annot['entities'])
    entities = spans_from_biluo_tags(doc, tags)
    doc.ents = entities
    docs_train.append(doc)
	
srsly.write_json(os.path.join(path_train, 'finaltrain.json'), [docs_to_json(docs_train)])
```
Then I can use the output file in the command line argument to train the model.

However after running my code in a new environment with Spacy 3.0.5, at the beginning of the code at the imports I get the following error

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-1-16069662e278> in <module>
      1 import spacy
      2 import srsly
----> 3 from spacy.gold import docs_to_json, biluo_tags_from_offsets, spans_from_biluo_tags

ModuleNotFoundError: No module named 'spacy.gold'
```

If `apacy.gold` has been removed what changes can I make either to the code above, or to my data set format to get it working using the command line training?

**EDIT**: I see in the documentation that you have converted to an `Example` object, 

```
# Training data for an entity recognizer (option 2)
doc = nlp("Laura flew to Silicon Valley.")
gold_dict = {"entities": [(0, 5, "PERSON"), (14, 28, "LOC")]}
example = Example.from_dict(doc, gold_dict)
```

although it doesn't say it, it seems that a training  instance is considered to be the same as a sentence? Or does it represent the entire document with multiple sentences?

If the example provided only shows a single sentence of the document - how do we add more than one sentence?

**Comment by svlandeg** [comment]
This is a good question for the Discussions forum, so I'll move it there. The original issue will be closed and will link to the new open thread on the forum.

**Comment by polm** [comment, marked as answer]
Managing training data has generally gotten simpler in v3, so all you need to do is create docs with the correct data on them (in your case the correct entities) and save them as a DocBin. The [sample projects](https://github.com/explosion/projects/tree/v3/tutorials) should be a good reference for this. 

I answered a similar question recently [here](https://github.com/explosion/spaCy/discussions/7555#discussioncomment-542044). Here's the sample code for that:

```
from spacy.training import Example
from spacy.tokens import DocBin

TRAINING_DATA = ... # a list of your training data instances
db = DocBin()
for text, gold_dict in TRAINING_DATA:
    example = Example.from_dict(nlp.make_doc(text), gold_dict)
    db.add(example.reference)

db.to_disk("train.spacy")
```

> although it doesn't say it, it seems that a training instance is considered to be the same as a sentence? Or does it represent the entire document with multiple sentences?
> 
> If the example provided only shows a single sentence of the document - how do we add more than one sentence?

I don't understand this question. Each training instance is a Doc, which can contain multiple sentences. In your old training format, each value in the top-level array would be converted into a Doc. Does that answer your question, or can you clarify what you meant?

**Reply by erotavlas** [reply]
I still don't understand  - In the example what is the value of `TRAINING_DATA`, and what are the values of `text` and `gold_dict` in the for loop?  Maybe knowing those might help.

**Reply by polm** [reply]
`TRAINING_DATA` is the example training data you gave:

    [("""Run date: 11/12/15 1705 """,{'entities':[(10,18,'DATE')]}),("""NAME: LNAMEZ,FNAME DOB: 12/12/1901 """,{'entities':[(6,18,'PERSON'),(36,46,'DATE')]})]

So in the first run of the for loop the text is `Run date: 11/12/15 1705` and `gold_dict` is `{'entities':[(10,18,'DATE')]}`. 

Does that make sense?

**Reply by erotavlas** [reply]
Yes thank you.
